File: src/main/java/rag/preprocess/TextNormalizer.java
package rag.preprocess;

import java.util.Arrays;
import java.util.List;

public class TextNormalizer {

    public String normalize(String text) {
        return text
                .replace("İ", "i").replace("I", "ı")
                .toLowerCase()
                .replaceAll("[^a-z0-9ğüşöçı ]", " ")
                .replaceAll("\\s+", " ")
                .trim();
    }

    public List<String> tokenize(String text) {
        return Arrays.asList(normalize(text).split(" "));
    }
}

File: src/main/python/rag/preprocess/text_normalizer.py
from typing import List
import re


class TextNormalizer:
    def normalize(self, text: str) -> str:
        normalized: str = (
            text.replace("İ", "i")
            .replace("I", "ı")
            .lower()
        )
        normalized = re.sub(r'[^a-z0-9ğüşöçıİ ]', ' ', normalized)
        normalized = re.sub(r'\s+', ' ', normalized)
        return normalized.strip()

    def tokenize(self, text: str) -> List[str]:
        return self.normalize(text).split()

